Mathlib/Algebra/GroupWithZero/Range.lean

```diff
@@ -111,8 +111,55 @@ lemma mem_valueGroup {b : Bˣ} (hb : b.1 ∈ range f) : b ∈ valueGroup f := by
 lemma inv_mem_valueGroup {b : Bˣ} (hb : b.1 ∈ range f) : b⁻¹ ∈ valueGroup f :=
   Subgroup.inv_mem _ (mem_valueGroup f hb)
 
-end MonoidWithZero
+end MonoidWithZeroHom
+
+noncomputable section Restrict
+
+variable [MonoidWithZero A] [GroupWithZero B] [MonoidWithZeroHomClass F A B] {f}
+
+open Classical in
+/-- The inclusion of `valueGroup₀ f` into `B` as a multiplicative homomorphism. -/
+def valueGroup₀_MulWithZeroEmbedding : valueGroup₀ f →*₀ B :=
+  MonoidWithZeroHom.comp (WithZero.withZeroUnitsEquiv (G := B))
+    <| WithZero.map' (valueGroup f).subtype
 
+
+variable (f) in
+open Classical in
+/-- This is the restriction of `f` as a function taking values in `valueGroup₀ f`. It cannot land
+in `valueMonoid₀ f` because in general `f a` needs not be a unit, so it will not be in
+`valueMonoid₀ f`. -/
+@[simps!]
+def restrict₀ : A →*₀ (valueGroup₀ f) where
+  toFun a :=
+    if h : f a ≠ 0 then (⟨Units.mk0 (f a) h, mem_valueGroup _ ⟨a, rfl⟩⟩ : valueGroup f) else 0
+  map_one' := by simp; rfl
+  map_mul' := by
+    intro a b
+    simp only [map_mul, ne_eq, Units.mk0_mul, dite_mul, zero_mul]
+    split_ifs with h hb ha
+    any_goals rfl
+    all_goals rw [mul_eq_zero] at h; tauto
+  map_zero' := by simp
+
+lemma restrict₀_of_ne_zero {a : A} (h : f a ≠ 0) :
+    restrict₀ f a = (⟨Units.mk0 (f a) h, mem_valueGroup _ ⟨a, rfl⟩⟩ : valueGroup f) :=
+  by simp [restrict₀, h]
+
+lemma restrict₀_of_eq_zero {a : A} (h : f a = 0) :
+    restrict₀ f a = 0 := by simp [restrict₀, h]
+
+lemma restrict₀_eq_zero_iff {a : A} : restrict₀ f a = 0 ↔ f a = 0 := by
+  refine ⟨fun h ↦ ?_, restrict₀_of_eq_zero⟩
+  · by_contra H; simp [H] at h
+
+lemma restrict₀_eq (a : A) : valueGroup₀_MulWithZeroEmbedding (restrict₀ f a) = f a := by
+  simp [restrict₀]
+  split_ifs with h
+  · simp [h]
+  · rfl
+
```

Comment:
```suggestion
lemma restrict₀_of_ne_zero {a : A} (h : f a ≠ 0) :
    restrict₀ f a = (⟨Units.mk0 (f a) h, mem_valueGroup _ ⟨a, rfl⟩⟩ : valueGroup f) := by
  simp [h]

lemma restrict₀_of_eq_zero {a : A} (h : f a = 0) :
    restrict₀ f a = 0 := by
  simp [h]

lemma restrict₀_eq_zero_iff {a : A} : restrict₀ f a = 0 ↔ f a = 0 := by
  simp

lemma restrict₀_eq (a : A) : valueGroup₀_MulWithZeroEmbedding (restrict₀ f a) = f a := by
  rw [restrict₀_apply]
  split_ifs <;>
  simp_all

```
since you tagged the hom with `simps`, we don't need to mention in in the simp set

and the valueGroup (not zero) lemma above helps avoid the heavy rfl

Reply:
Nice, implemented!